### DISCUSSION TITLE
Out of memory issues on large imports

### DISCUSSION THREAD

**Discussion by sxsws** [directus/directus#13299] [Archive]
Importing approx 1.1 million rows of a fairly simple collection and directus is exiting with out of memory errors and a dump of some of the garbage collection errors.

I assume it's trying to read the whole file in one go then import it as no records are actually imported when it crashes.

If so, it's going to need to batch the import up into x records at a time, commit and repeat to avoid running out of memory

### Errors Shown

```
<--- Last few GCs --->

[4339:0x562196a1e0f0]   487496 ms: Scavenge 1933.1 (1946.0) -> 1932.5 (1946.5) MB, 11.2 / 0.0 ms  (average mu = 0.011, current mu = 0.000) allocation failure 
[4339:0x562196a1e0f0]   489976 ms: Mark-sweep 1933.3 (1946.5) -> 1932.6 (1946.5) MB, 2473.0 / 0.0 ms  (average mu = 0.011, current mu = 0.010) allocation failure scavenge might not succeed


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x7fc817dc53b9]
Security context: 0x38880349b9a1 <JSObject>
    1: values [0x38880348cd61](this=0x302dd46c0759 <JSArray[1]>)
    2: StubFrame [pc: 0x7fc817d8751a]
    3: StubFrame [pc: 0x7fc817d45c40]
    4: new DateHelper(aka DateHelper) [0x8f606799151] [/home/a/directus/directus_test2/node_modules/directus/dist/database/helpers/date/types.js:~6] [pc=0x1e912694928](this=0x3a24b1b40591 <the_hole>)
    5: arguments ad...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x7fc8171165fc node::Abort() [/lib/x86_64-linux-gnu/libnode.so.72]
 2: 0x7fc817046fdc  [/lib/x86_64-linux-gnu/libnode.so.72]
 3: 0x7fc8172d7a2a v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/lib/x86_64-linux-gnu/libnode.so.72]
 4: 0x7fc8172d7de4 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/lib/x86_64-linux-gnu/libnode.so.72]
 5: 0x7fc81748a259  [/lib/x86_64-linux-gnu/libnode.so.72]
 6: 0x7fc81749ab16 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/lib/x86_64-linux-gnu/libnode.so.72]
 7: 0x7fc81749b95e v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/lib/x86_64-linux-gnu/libnode.so.72]
 8: 0x7fc81749dcfd v8::internal::Heap::AllocateRawWithLightRetry(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/lib/x86_64-linux-gnu/libnode.so.72]
 9: 0x7fc81749dd69 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/lib/x86_64-linux-gnu/libnode.so.72]
10: 0x7fc817462ab3 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/lib/x86_64-linux-gnu/libnode.so.72]
11: 0x7fc8177b8277 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/lib/x86_64-linux-gnu/libnode.so.72]
12: 0x7fc817dc53b9  [/lib/x86_64-linux-gnu/libnode.so.72]
Aborted (core dumped)
npm ERR! code 134
npm ERR! path /home/a/directus/directus_test2
npm ERR! command failed
npm ERR! command sh -c directus "start"

```

**Comment by azrikahar** [comment]
This is technically a known constraint as mentioned here: https://github.com/directus/directus/discussions/7246#discussioncomment-1870098, and there are also internal & public discussion such as #11274 regarding improvements to the Import experience as a whole, so I'll move this to discussion for now.

Some of the additional reasons why large imports can be taxing in terms of resource is that there are also things like revision tracking, user_created injection, relational traversing and so on, _for every row_. A queue to process them in batches would definitely be much better than the current behavior.

Another interesting point brought up before in the past is another type of import that does not involve activity, revision etc to cut down on the extra operations and improve the import performance, but limited to admins only! 

> If so, it's going to need to batch the import up into x records at a time, commit and repeat to avoid running out of memory

Very true! #12201 has implemented the same batching/async nature to "export" that we'd like for "import" as well, so we may be able to lift some of the logic from there.

**Reply by sxsws** [reply]
Thanks for the response!
- I think limiting the import to just admins would be fine especially given that not everything is validated.
- If you need a queue manager, I  don't think adding a hard dependency like redis is unreasonable given that it would help future features like managing jobs/handlers in data flows and would allow scheduled jobs that storing them in the memory isn't really a viable alternative to.
- as for activity tracking, I tend to keep it switched off. Why is it on by default? I can imagine only a small percentage of collections need such tracking.
- Also, may I suggest that in the import section there is some detail about the headers needed for the data. I find myself creating dummy data and exporting the data to see which headers to use, so a foldout panel of the names of the headers would be extremely useful. Especially so for relations as the db field may be called `foos` but the header has to be `foos.id` which breaks from the `the header is the name of the field` convention.

I don't think this is a current deal breaker with some db admin software being able to import from csv themselves (though I'm yet to find one that can import geometry like point!)
